File: PropertyTestsFs/PotterDictionaryPropertyTests.fs
﻿module PotterDictionaryPropertyTests

open FsCheck
open FsCheck.NUnit
open FsCheck.Fluent
open System.Linq
open Code

let unitPrice = PotterCalculator.UnitPrice

let checkPriceOfBooks (titles:string[]) (expectedPrice:decimal) =
    let actualPrice = PotterCalculator.CalculatePrice titles
    actualPrice = expectedPrice

let checkPriceOfBooksSpec (gen:Gen<string[]>) (expectedPriceFun:string[] -> decimal) =
    let dontShrinkStringArrays = new System.Func<string[], seq<string[]>>(fun _ -> Seq.empty) 
    let specBuilder = Spec.For (gen, fun titles -> checkPriceOfBooks titles (expectedPriceFun titles))
    specBuilder
        .Shrink(dontShrinkStringArrays)
        .QuickCheckThrowOnFailure()

let checkPriceOfBooksSpecWithFixedPrice (gen:Gen<string[]>) (expectedPrice:decimal) =
    checkPriceOfBooksSpec gen (fun _ -> expectedPrice)

let titlesAreDifferent titles =
    Seq.length(Seq.distinct titles) = Seq.length titles

let combinationsOfTwoDifferentTitles all =
    seq {
        for t1 in all do
        for t2 in all do
        let titles = [|t1; t2|]
        if titlesAreDifferent(titles) then yield titles
    }

let combinationsOfThreeDifferentTitles all =
    seq {
        for t1 in all do
        for t2 in all do
        for t3 in all do
        let titles = [|t1; t2; t3|]
        if titlesAreDifferent(titles) then yield titles
    }

let combinationsOfFourDifferentTitles all =
    seq {
        for t1 in all do
        for t2 in all do
        for t3 in all do
        for t4 in all do
        let titles = [|t1; t2; t3; t4|]
        if titlesAreDifferent(titles) then yield titles
    }

let combinationsOfFiveDifferentTitles all =
    seq {
        for t1 in all do
        for t2 in all do
        for t3 in all do
        for t4 in all do
        for t5 in all do
        let titles = [|t1; t2; t3; t4; t5|]
        if titlesAreDifferent(titles) then yield titles
    }

let genOneTitle = Gen.elements(HarryPotterBooks.Titles)
let genOneTitleArray = Gen.map (fun title -> Seq.singleton title |> Seq.toArray) genOneTitle
let genTwoDifferentTitles = Gen.elements(combinationsOfTwoDifferentTitles(HarryPotterBooks.Titles))
let genThreeDifferentTitles = Gen.elements(combinationsOfThreeDifferentTitles(HarryPotterBooks.Titles))
let genFourDifferentTitles = Gen.elements(combinationsOfFourDifferentTitles(HarryPotterBooks.Titles))
let genFiveDifferentTitles = Gen.elements(combinationsOfFiveDifferentTitles(HarryPotterBooks.Titles))

let genMultipleTitlesTheSame =
    gen {
        let! title = genOneTitle
        let! n = Gen.choose(0, 20)
        return Enumerable.Repeat(title, n) |> Seq.toArray
    }

let genOverlappingFourDifferentTitlesPlusTwoDifferentTitles =
    gen {
        let! four = genFourDifferentTitles
        let! two = Gen.elements(combinationsOfTwoDifferentTitles(four))
        return Array.append four two
    }

[<Property>]
let ``one book``() = 
    let expectedPrice = unitPrice
    checkPriceOfBooksSpecWithFixedPrice genOneTitleArray expectedPrice

[<Property>]
let ``multiple books the same``() = 
    let expectedPriceFun = fun titles -> (Seq.length titles |> decimal) * unitPrice
    checkPriceOfBooksSpec genMultipleTitlesTheSame expectedPriceFun

[<Property>]
let ``two books different``() = 
    let expectedPrice = 2m * unitPrice * 0.95m
    checkPriceOfBooksSpecWithFixedPrice genTwoDifferentTitles expectedPrice

[<Property>]
let ``three books different``() = 
    let expectedPrice = 3m * unitPrice * 0.90m
    checkPriceOfBooksSpecWithFixedPrice genThreeDifferentTitles expectedPrice

[<Property>]
let ``four books different``() = 
    let expectedPrice = 4m * unitPrice * 0.80m
    checkPriceOfBooksSpecWithFixedPrice genFourDifferentTitles expectedPrice

[<Property>]
let ``five books different``() = 
    let expectedPrice = 5m * unitPrice * 0.75m
    checkPriceOfBooksSpecWithFixedPrice genFiveDifferentTitles expectedPrice

[<Property>]
let ``overlapping four different books plus two different books``() =
    let gen = genOverlappingFourDifferentTitlesPlusTwoDifferentTitles
    let expectedPrice = (4m * unitPrice * 0.80m) + (2m * unitPrice * 0.95m)
    checkPriceOfBooksSpecWithFixedPrice gen expectedPrice

File: PropertyTestsCs/PotterDictionaryPropertyTests.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using Code;
using FsCheck;
using FsCheck.Fluent;
using Microsoft.FSharp.Core;

namespace PropertyTests
{
    internal class PotterDictionaryPropertyTests
    {
        private const decimal UnitPrice = PotterCalculator.UnitPrice;

        private static bool CheckPriceOfBooks(string[] titles, decimal expectedPrice)
        {
            var actualPrice = PotterCalculator.CalculatePrice(titles);
            return actualPrice == expectedPrice;
        }

        private static void CheckPriceOfBooksSpec(Gen<string[]> gen, Func<string[], decimal> expectedPriceFunc)
        {
            Spec
                .For(gen, titles => CheckPriceOfBooks(titles, expectedPriceFunc(titles)))
                .Shrink(_ => Enumerable.Empty<string[]>())
                .QuickCheckThrowOnFailure();
        }

        private static void CheckPriceOfBooksSpecWithFixedPrice(Gen<string[]> gen, decimal expectedPrice)
        {
            CheckPriceOfBooksSpec(gen, _ => expectedPrice);
        }

        private static bool TitlesAreDifferent(ICollection<string> titles)
        {
            return titles.Distinct().Count() == titles.Count;
        }

        private static IEnumerable<string[]> CombinationsOfTwoDifferentTitles(string[] allTitles)
        {
            return from title1 in allTitles
                   from title2 in allTitles
                   let titles = new[] { title1, title2 }
                   where TitlesAreDifferent(titles)
                   select titles;
        }

        private static IEnumerable<string[]> CombinationsOfThreeDifferentTitles(string[] allTitles)
        {
            return from title1 in allTitles
                   from title2 in allTitles
                   from title3 in allTitles
                   let titles = new[] { title1, title2, title3 }
                   where TitlesAreDifferent(titles)
                   select titles;
        }

        private static IEnumerable<string[]> CombinationsOfFourDifferentTitles(string[] allTitles)
        {
            return from title1 in allTitles
                   from title2 in allTitles
                   from title3 in allTitles
                   from title4 in allTitles
                   let titles = new[] { title1, title2, title3, title4 }
                   where TitlesAreDifferent(titles)
                   select titles;
        }

        private static IEnumerable<string[]> CombinationsOfFiveDifferentTitles(string[] allTitles)
        {
            return from title1 in allTitles
                   from title2 in allTitles
                   from title3 in allTitles
                   from title4 in allTitles
                   from title5 in allTitles
                   let titles = new[] { title1, title2, title3, title4, title5 }
                   where TitlesAreDifferent(titles)
                   select titles;
        }

        private static readonly Gen<string> GenOneTitle = Gen.elements(HarryPotterBooks.Titles);
        private static readonly FSharpFunc<string, string[]> StringToSingletonArray = FSharpFunc<string, string[]>.FromConverter(s => new[] {s});
        private static readonly Gen<string[]> GenOneTitleArray = Gen.map(StringToSingletonArray, GenOneTitle);
        private static readonly Gen<string[]> GenTwoDifferentTitles = Gen.elements(CombinationsOfTwoDifferentTitles(HarryPotterBooks.Titles));
        private static readonly Gen<string[]> GenThreeDifferentTitles = Gen.elements(CombinationsOfThreeDifferentTitles(HarryPotterBooks.Titles));
        private static readonly Gen<string[]> GenFourDifferentTitles = Gen.elements(CombinationsOfFourDifferentTitles(HarryPotterBooks.Titles));
        private static readonly Gen<string[]> GenFiveDifferentTitles = Gen.elements(CombinationsOfFiveDifferentTitles(HarryPotterBooks.Titles));

        private static readonly Gen<string[]> GenMultipleTitlesTheSame =
            from title in GenOneTitle
            from n in Gen.choose(0, 20)
            select Enumerable.Repeat(title, n).ToArray();

        private static readonly Gen<string[]> GenOverlappingFourDifferentTitlesPlusTwoDifferentTitles =
            from four in GenFourDifferentTitles
            from two in Gen.elements(CombinationsOfTwoDifferentTitles(four))
            select four.Concat(two).ToArray();

        [FsCheck.NUnit.Property]
        public void OneBook()
        {
            const decimal expectedPrice = UnitPrice;
            CheckPriceOfBooksSpecWithFixedPrice(GenOneTitleArray, expectedPrice);
        }

        [FsCheck.NUnit.Property]
        public void MultipleBooksTheSame()
        {
            CheckPriceOfBooksSpec(GenMultipleTitlesTheSame, titles => titles.Length * UnitPrice);
        }

        [FsCheck.NUnit.Property]
        public void TwoDifferentBooks()
        {
            const decimal expectedPrice = 2 * UnitPrice * 0.95m;
            CheckPriceOfBooksSpecWithFixedPrice(GenTwoDifferentTitles, expectedPrice);
        }

        [FsCheck.NUnit.Property]
        public void ThreeDifferentBooks()
        {
            const decimal expectedPrice = 3 * UnitPrice * 0.90m;
            CheckPriceOfBooksSpecWithFixedPrice(GenThreeDifferentTitles, expectedPrice);
        }

        [FsCheck.NUnit.Property]
        public void FourDifferentBooks()
        {
            const decimal expectedPrice = 4 * UnitPrice * 0.80m;
            CheckPriceOfBooksSpecWithFixedPrice(GenFourDifferentTitles, expectedPrice);
        }

        [FsCheck.NUnit.Property]
        public void FiveDifferentBooks()
        {
            const decimal expectedPrice = 5 * UnitPrice * 0.75m;
            CheckPriceOfBooksSpecWithFixedPrice(GenFiveDifferentTitles, expectedPrice);
        }

        [FsCheck.NUnit.Property]
        public void OverlappingFourDifferentBooksPlusTwoDifferentBooks()
        {
            var gen = GenOverlappingFourDifferentTitlesPlusTwoDifferentTitles;
            const decimal expectedPrice = (4 * UnitPrice * 0.80m) + (2 * UnitPrice * 0.95m);
            CheckPriceOfBooksSpecWithFixedPrice(gen, expectedPrice);
        }
    }
}